HPCC-16833 Regression suite mode to fail on row leaks and missing stops

richardkchapman opened the pull request:
Designed for use from regression suite, option to treat row leaks as fatal, to
try to make sure they are not accidentally introduced.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-16833 Regression suite mode to fail on row leaks and missing stops

Designed for use from regression suite, option to treat row leaks as fatal, to
try to make sure they are not accidentally introduced.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman commented:
@ghalliday Pleae review

HPCCSmoketest commented:
Automated Smoketest
Sha: fd9760cfaa183ebfde0041662b78202997de0df2
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.731 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@AttilaVamos what scripts need to be changed to ensure that this option is used when we run the regression suite?  Should it be a default option within ecl-test?

AttilaVamos commented:
@ghalliday I talked about it with Richard yesterday. I will create a new JIRA (today) to add this capability to Regression Test Engine. I presume there can be some more parameters to control engines so my plan is to store them into ecl-test.json config file to avoid code modification every time when we want to modify/extend the engines behaviour.

richardkchapman committed:
HPCC-16833 Regression suite mode to fail on row leaks and missing stops

Address some false positives.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest
Sha: 8c7747ab246013bc882ae7b6758e3ea4a114549f
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.558 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman committed:
HPCC-16849 N-Way input activity reports RESET without STOP

Stop once started - there is no further processing by the activity itself.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-16848 QUANTILE activity is not releasing rows to roxie heap

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest
Sha: 267b1e7f6c3541a727e699379a7558f5a04baeef
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.937 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman committed:
HPCC-16852 Limited prefix joins may leak roxie rows

This seems to have already been fixed in the Thor code this was originally
adapted from.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest
Sha: 3ff959cf4496fc15ff96433c17f15f800713bffe
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.172 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@ghalliday I think the commits here fix the existing failures that this mode causes in the Roxie regression suite - so it should be safe to merge now. The new Roxie flag is pretty safe (as no-one will use it outside of the regression suite) but the changes to fix the leaks are potentially more dangerous!

ghalliday commented on roxie/ccd/ccdserver.cpp:
```diff
@@ -16283,14 +16278,11 @@ class CRoxieServerNWayInputActivity : public CRoxieServerNWayInputBaseActivity
         }
-
         // NB: Whatever pulls this nwayinput activity, starts and stops the selectedInputs and selectedJunctions
+        // But the N-Way input itself is now donw with processing, and can stop itself/dependencies.
```
typo donw->done
